File: Lecture023 2D arrays/wavePrint.cpp
#include<iostream>
#include<vector>
using namespace std;
vector<int> wavePrint(vector<vector<int> > arr, int nRows, int mCols)
{
    vector<int> ans;
    
    for(int col=0; col<mCols; col++) {
        
        if( col&1 ) {
            //odd Index -> Bottom to top
            
            for(int row = nRows-1; row>=0; row--) {
                //cout << arr[row][col] <<" ";
                ans.push_back(arr[row][col]);
            }    
        }
        else
        {
            // 0 or even iondex -> top to bottom
            for(int row = 0; row<nRows; row++) {
                //cout << arr[row][col] << " ";
                ans.push_back(arr[row][col]);
            }
        }   
    }
    return ans;
}

File: 023_2D_arrays/WavePrint.java
public class Solution {
    public static int[] wavePrint(int arr[][], int nRows, int mCols) {
        // Write your code here.
        int[] ans = new int[nRows * mCols]; // array to store the values of wave and return

        int idx = 0; // index for traversing the ans indexes

        // Here i represents columns & j represents Rows
        for(int i = 0; i < mCols; i++){

            if(i % 2 != 0){ // bottom to top

                for(int j = nRows - 1; j >= 0; j--){
                    ans[idx++] = arr[j][i];
                }
            }
            else { // top to bottom
                for(int j = 0; j < nRows; j++){
                    ans[idx++] = arr[j][i];
                }
            }
        }

        return ans;
    }
}
